npm refactoring

jmlemetayer opened the pull request:
The patchset refactors the npm support in Yocto:
 - The npm fetcher downloads only the package source (and not the dependencies,
   like the other fetchers) (patch submitted in the bitbake-devel list).
 - The npm class handles the dependencies with npm (and not manually).
 - The npm recipe creation is simplified to avoid issues.

This patchset may remove some features (lockdown file, license management for
dependencies) but fixes the majority of the npm issues. All of these issues
from the bugzilla.yoctoproject.org are resolved by this patchset:
#10237, #10760, #11028, #11728, #11902, #12534

The fetcher and recipetool are now aware of a 'latest' keyword for the version
which allow to build the latest version available on the registry. This feature
fixes the two issues: #10515, #11029

Moreover, even if I cannot test the issue #13415, I think it should be fixed.

I have tested the recipe creation and builds using a self made example made to
generate build failures. It can be found here:
 - https://github.com/savoirfairelinux/node-server-example
 - https://npmjs.com/package/@savoirfairelinux/node-server-example

The test steps are these ones:
  $ source poky/oe-init-build-env
  $ bitbake-layers add-layer ../meta-openembedded/meta-oe
  $ devtool add "npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=latest"
  $ devtool build savoirfairelinux-node-server-example
  $ bitbake-layers create-layer ../meta-test
  $ bitbake-layers add-layer ../meta-test
  $ devtool finish savoirfairelinux-node-server-example ../meta-test
  $ echo IMAGE_INSTALL_append = '" savoirfairelinux-node-server-example"' >> conf/local.conf
  $ bitbake core-image-minimal

And the 'devtool add' urls could be one of these:
 - npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=latest
 - npm://registry.npmjs.org;name=@savoirfairelinux/node-server-example;version=1.0.0
 - git://github.com/savoirfairelinux/node-server-example.git;protocol=https

I also tested with angular applications with huge dependencies and it worked
very well. When this patchset will be merged I will be able to submit a full
angular application support in Yocto.

Commit:
npm.bbclass: refactor the npm class

Many issues were related to npm dependencies badly handled: package
names, installation directories, ... In fact npm is using an install
algorithm [1] which is hard to reproduce / anticipate. Moreover some
npm packages use scopes [2] which adds more complexity.

The simplest solution is to let npm do its job. Assuming the fetcher
only get the sources of the package, the class will now run
'npm install' to create a build directory. The build directory is then
copied wisely to the destination.

1: https://docs.npmjs.com/cli/install#algorithm
2: https://docs.npmjs.com/about-scopes

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

Commit:
fetch2/npm.py: refactor the npm fetcher

This commit refactors the npm fetcher to improve some points and fix
others:

 - The big change is that the fetcher is only fetching the package
   source and no more the dependencies. Thus the npm fetcher act as the
   other fetchers e.g git, wget. The dependencies will now be fetched
   by the npm class.

 - This commit also fixes a lot of issues with the package names (exotic
   characters, scoped packages) which were badly handled.

 - The validation files - lockdown.json and npm-shrinkwrap.json - are no
   longer used by the fetcher. Instead, the downloaded tarball is
   validated with the 'integrity' and 'shasum' provided in the npm view
   of the package [1] [2].

 - The lockdown file had generation issues and is no longer relevant
   with the latest shrinkwrap files. The shrinkwrap file is now used
   by the npm class.

1: https://docs.npmjs.com/files/package-lock.json#integrity
2: https://w3c.github.io/webappsec/specs/subresourceintegrity

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

Commit:
devtool: update command line options for npm

This commit renames the '--fetch-dev' option into '--npm-dev' which is
more easily understandable.

It also adds the '--npm-registry' option to allow creating a npm recipe
with a non default npm registry (e.g. if the SRC_URI is using git://).

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

Commit:
devtool/standard.py: update the append file for the npm recipes

When creating a recipe using devtool, a workspace is created to store
the new recipe, the recipe source and some append files. These append
files are used by devtool to build the recipe using externalsrc (to use
the source which are in the workspace). They can also have some
additional actions according to the class of the recipe.

This commit updates the append file for the npm recipes. The
devtool / externalsrc files are removed in the npm build directory
instead of the install directory.

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

Commit:
recipetool/create_npm.py: refactor the npm recipe creation handler

This commit refactors the npm recipe creation handler to use the new npm
behavior. The process is kept as simple as possible and only generates
the shrinkwrap file.

To avoid naming issues the recipe name is now extracted from the npm
package name and not directly map.

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

gdoffe commented on bitbake/lib/bb/fetch2/npm.py:
```diff
@@ -2,300 +2,200 @@
     def supports(self, ud, d):
         """
-        Check to see if a given url can be fetched with npm
+            Check to see if a given url can be fetched with npm
```
remove 'to see'

gdoffe commented on bitbake/lib/bb/fetch2/npm.py:
```diff
@@ -2,300 +2,200 @@
     def urldata_init(self, ud, d):
         """
-        init NPM specific variable within url data
+            Init npm specific variable within url data
```
variables ?

gdoffe commented on bitbake/lib/bb/fetch2/npm.py:
```diff
@@ -2,300 +2,200 @@
+        #
+        # This trick can be useful with devtool / recipetool to automatically
+        # fetch the latest version, BUT these files (the ones with non-existent
+        # versions) must be updated every times.
```
If I understand well, it is not  related to 'latest' keyword it could be anything else ?
If its the case it is dangerous, the described behavior should only be applied in case of 'latest' version specified.

jmlemetayer replied:
Your right. I will update the behavior to accept only existent version or the "latest" keyword.

gdoffe commented on bitbake/lib/bb/fetch2/npm.py:
```diff
@@ -2,300 +2,200 @@
         """
-        Runs npm pack on a full package name.
-        Returns the filename of the downloaded package
+            Run the 'wget' command with progress
```
'download progress status' you mean ?

gdoffe commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
+
+npm_install_shrinkwrap() {
+    # This function ensures that there is a shrinkwrap file in the specified
+    # directory. A shrinkwrap file is mandatory to be able to have reproducible
```
remove 'to be able'

gdoffe commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
+    # This function ensures that there is a shrinkwrap file in the specified
+    # directory. A shrinkwrap file is mandatory to be able to have reproducible
+    # builds. If the shrinkwrap file is not already included in the sources,
+    # the recipe can provides one by using the NPM_SHRINKWRAP option.
```
provide

gdoffe commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
+        cp --no-preserve=ownership --recursive ${B}/bin/. ${D}/${bindir}
+    fi
+
+    # If the package (or his dependencies) uses node-gyp to build native addons,
```
'its' not 'his'

gdoffe reviewed (changes requested):
**devtool/standard.py: update the append file for the npm recipes:**
What is the 'append file' and in which context is this file used ?
I guess it is related to an yocto recipe, but it could be useful to mention it.

gdoffe reviewed (commented):
**fetch2/npm.py: refactor the npm fetcher:**
That's a big commit with new features, bug fixes, etc...
Is this possible to split thta commit in something more atomic and precise ?

gdoffe reviewed (commented):
**npm.bbclass: refactor the npm class:**
I would replace:
"The simplest solution is to keep it as simple as possible and to let
npm do its job."
by:
"The simplest solution is to let npm do its job."

gdoffe reviewed (changes requested):
**recipetool/create_npm.py: refactor the npm recipe creation handler:**
generates

Commit:
recipetool/create.py: replace 'latest' keyword for npm

The new npm fetcher allows the 'latest' keyword to be used to download
the latest version on the registry. But the keyword must be replace
as soon as the version is determined to have a stable generated recipe.

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

jmlemetayer commented:
> **devtool/standard.py: update the append file for the npm recipes:**
> What is the 'append file' and in which context is this file used ?
> I guess it is related to an yocto recipe, but it could be useful to mention it.

Done

jmlemetayer commented:
> The simplest solution is to let npm do its job

Done

jmlemetayer commented:
> **recipetool/create_npm.py: refactor the npm recipe creation handler:**
> generates

Done

Commit:
recipetool/create.py: remove the 'noverify' url parameter

This commit removes the 'noverify' parameter which was added to the url
to fix warnings with the shrinkwrap / lockdown file generation. This is
not needed anymore with the new npm fetcher.

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

jmlemetayer commented:
> **fetch2/npm.py: refactor the npm fetcher:**
> That's a big commit with new features, bug fixes, etc...
> Is this possible to split thta commit in something more atomic and precise ?

The refactoring is done on 3 files: the fetcher, the class and the recipetool handler. This is very hard to split the diff on these files.
However I split mandatory diff on the other files as atomic as possible.

eloi29 commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
 
-# function maps arch names to npm arch names
-def npm_oe_arch_map(target_arch, d):
+def npm_arch(target_arch):
```
You are renaming the function and removing the function description.
I would do that in a seperate commit

eloi29 commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
 }
 
-FILES_${PN} += " \
+FILES_${PN} = " \
```
I would explain in a separate commit why you are using 
`FILES_${PN} = " ` instead of `FILES_${PN} += " `

eloi29 commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
     ${bindir} \
-    ${libdir}/node \
-    ${NPM_INSTALLDIR} \
+    ${libdir} \
```
I would explain in a separate commit why:
- renaming ${libdir}/node 
- removing ${NPM_INSTALLDIR}

eloi29 commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
+
+    # The npm command use by default a cache which is located in '~/.npm'. In
+    # order to force the next npm commands to disable caching, the npm cache
+    # needs to be cleared. But to not mess with the local cache, the npm
```
`but not to mess instead `of `but to not mess`

eloi29 replied:
In addition, the verb "mess" looks to be informal. I would use an other word

eloi29 commented on meta/classes/npm.bbclass:
```diff
@@ -1,94 +1,171 @@
+    # order to force the next npm commands to disable caching, the npm cache
+    # needs to be cleared. But to not mess with the local cache, the npm
+    # config needs to be updated to use another cache directory. And to not
+    # mess with the local '~/.npmrc` file, the HOME needs to be updated as well.
```
I would reword like below:

the HOME needs to be updated as well to avoid conflicts with the local '~/.npmrc' file.

eloi29 reviewed (commented):
> 
> This commit refactors the npm fetcher to improve some points and fix
> others:
I would name it. What are you improving ? What are you fixing ?

> 
> The big change is that the fetcher is only fetching the package source
> and no more the dependencies. This is to act as the other fetchers
> (git, wget, ...). The dependencies will now be fetched by the npm class.
> 

I would reword:

` This is to act as the other fetchers (git, wget, ...) ` to `Thus the npm fetcher act as the other fetchers e.g git, wget`

> This commit also fixes a lot of issues with the package names (exotic
> characters, scoped packages) which were badly handled.

Have you got an exhaustive list of fixed issues ?

> 
> The validation files - lockdown.json and npm-shrinkwrap.json - are no
> more used by the fetcher. Instead, the downloaded tarball is validated
> with the shasum provided in the npm view of the package. The shrinkwrap
> file is now used by the npm class.

Is the use of shasum the standard way outside Yocto ? If it does, mention it.

> 
> Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

eloi29 commented on scripts/lib/recipetool/create_npm.py:
```diff
@@ -1,321 +1,215 @@
+        """
 
-        if 'buildsystem' in handled:
+        if "buildsystem" in handled:
```
why using double quotes instead of simple quotes ?

eloi29 reviewed (changes requested):
> When creating a recipe using devtool, a workspace is created to store
> the new recipe, the recipe source and some append files. These append
> files are used by devtool to build the recipe using externalsrc (to use
> the source which are in the workspace). They can also have some
> additional action according to the class on the recipe.

I would reword `They can also have some additional action according to the class on the recipe.`
I don't understand what you mean.
Otherwise LGTM.
> 
> This commit updates the append file for the npm recipes. The
> devtool / externalsrc files are removed in the npm build directory
>

eloi29 reviewed (changes requested):
I would write `the "noverify" parameter` instead of `the noverify parameter`.

Commit:
bitbake: utils.py: add sha384_file and sha512_file

This commit adds the "sha384_file" and "sha512_file" functions in order
to check the integrity of the downloaded npm packages as npm now use
Subresource Integrity:

  https://w3c.github.io/webappsec-subresource-integrity

Signed-off-by: Jean-Marie LEMETAYER <jean-marie.lemetayer@savoirfairelinux.com>

jmlemetayer commented:
> > This commit refactors the npm fetcher to improve some points and fix
> > others:
> > I would name it. What are you improving ? What are you fixing ?
> 
> > The big change is that the fetcher is only fetching the package source
> > and no more the dependencies. This is to act as the other fetchers
> > (git, wget, ...). The dependencies will now be fetched by the npm class.
> 
> I would reword:
> 
> `This is to act as the other fetchers (git, wget, ...)` to `Thus the npm fetcher act as the other fetchers e.g git, wget`
> 
> > This commit also fixes a lot of issues with the package names (exotic
> > characters, scoped packages) which were badly handled.
> 
> Have you got an exhaustive list of fixed issues ?
> 
> > The validation files - lockdown.json and npm-shrinkwrap.json - are no
> > more used by the fetcher. Instead, the downloaded tarball is validated
> > with the shasum provided in the npm view of the package. The shrinkwrap
> > file is now used by the npm class.
> 
> Is the use of shasum the standard way outside Yocto ? If it does, mention it.
> 
> > Signed-off-by: Jean-Marie LEMETAYER [jean-marie.lemetayer@savoirfairelinux.com](mailto:jean-marie.lemetayer@savoirfairelinux.com)

Done

jmlemetayer commented:
> > When creating a recipe using devtool, a workspace is created to store
> > the new recipe, the recipe source and some append files. These append
> > files are used by devtool to build the recipe using externalsrc (to use
> > the source which are in the workspace). They can also have some
> > additional action according to the class on the recipe.
> 
> I would reword `They can also have some additional action according to the class on the recipe.`
> I don't understand what you mean.
> Otherwise LGTM.
> 
> > This commit updates the append file for the npm recipes. The
> > devtool / externalsrc files are removed in the npm build directory

Done

jmlemetayer commented:
> I would write `the "noverify" parameter` instead of `the noverify parameter`.

Done
